Add Cudagraphs disable checking

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121033
* #121019
* __->__ #121018
* #121017

Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Add Cudagraphs disable checking

[ghstack-poisoned]

eellison committed:
Update on "Add Cudagraphs disable checking"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Cudagraphs disable checking"


Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Cudagraphs disable checking"


Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Cudagraphs disable checking"


Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Cudagraphs disable checking"


Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

ezyang commented on test/inductor/test_cudagraph_trees.py:
```diff
@@ -50,6 +52,13 @@
 
+def get_compile_fn(backend):
+    if backend == "cudagraphs":
+        return functools.partial(torch.compile, backend="cudagraphs")
```
Don't you also have an option to do inductor + cudagraphs (but not cudagraph trees?) or am I hallucinating

eellison replied:
That's basically old, legacy code. kept around because occasionally useful for debugging but not actively used

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -353,6 +353,8 @@ def default_debug_dir_root():
     "skipfiles_inline_module_allowlist",
 }
 
+cudagraph_backend_keep_input_mutation = False
```
needs docs

ezyang commented on torch/_inductor/cudagraph_utils.py:
```diff
@@ -21,6 +21,24 @@ def format_default_skip_message(reason: str) -> str:
+            break
+
+    if stack_trace:
+        msg = f"skipping cudagraphs due to mutaton on input. Found from : \n {stack_trace}"
```
mutation

ezyang commented on torch/_dynamo/backends/cudagraphs.py:
```diff
@@ -93,9 +101,11 @@ def meta_fk(meta):
     inputs = defaultdict(set)
     input_idx = 0
     mutated_inputs = set()
+    mutated_node = {}
```
Looks dead?

ezyang commented on torch/_dynamo/backends/cudagraphs.py:
```diff
@@ -134,25 +145,70 @@ def apply_cuda_graphs(gm):
+        fixed = count_tangents(aot_model)
+        if skip_msg := check_for_skip(aot_model, fixed):
+            perf_log.warning("skipping cudagraphs due to %s", skip_msg)
+            warnings.warn(skip_msg)
```
How are you deciding whether or not to warnings.warn here? For example, in the test in forwards_cudagraphs, you don't emit a warning.

eellison replied:
Yea, duplicate warning is a mistake

ezyang commented on torch/_dynamo/backends/cudagraphs.py:
```diff
@@ -134,25 +145,70 @@ def apply_cuda_graphs(gm):
+
+
+def check_for_skip(aot_model: torch.fx.GraphModule, num_fixed) -> Optional[str]:
+    if mut_skip := check_for_mutation(aot_model, num_fixed):
```
nit: empty string will also evaluate falsey. Better to do (...) is not None

ezyang reviewed (approved):
seems fine ig. I didn't carefully cross reference with cudagraph_trees. It does seem a little duplicatey

eellison committed:
Update on "Add Cudagraphs disable checking"


Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Cudagraphs disable checking"


Adds the same cudagraphs disable checking from inductor - cudagraph trees to cudagraphs backend.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
